api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * Optional interface for FileIO implementations to report metrics related
+ * to read/write operations.
+ */
+public interface FileIOMetrics extends Serializable {
+  String READ_BYTES = "read.bytes";
```

Comment:
I think one of the challenges might become the metric names and their usage with different systems. Prometheus for example uses suffixes like `_seconds` / `_bytes` / `_total` (https://prometheus.io/docs/practices/naming/). One option would we to have a metric name converter to match whatever the target metric system expects it to be.

Reply:
correct, that's where I was trying to get at. We would probably need some kind of naming conversion to happen to map to whatever system we'd want the metrics to be mapped to

Reply:
I'd definitely encourage you to take a look at the Flink metric implementation. It worked well for me to be able to use their metrics, plus my own metrics via the Prometheus metrics library, and merge them together. And it provides shims to various metric formats and is used for the same purpose (hadoop fs metrics): https://github.com/apache/flink/tree/master/flink-metrics

But I still agree that the simplest approach should be chosen initially to get the metrics working for S3FileIO. It wouldn't be hard to adapt them later and it's a highly requested feature.